### PR TITLE
[multistage] [bugfix] Disable shuffle rewrite

### PR DESCRIPTION
**Pull request by 61yao**
1) This can can be used to check how much performance improvement skip shuffle brings
2) It can also be used to quickly check whether skip shuffle introduces bugs

### CODE REVIEW

**Review by siddharthteotia — APPROVED** [review]

**Comment by siddharthteotia** [comment]
I am guessing we can use this to compare performance and even correctness between `broadcast join` , `shuffle join`,  `colocated partitioning` table join ? 

I guess pretty much any scenario where we force the plan to not have an exchange ?

Another question - are we using `shuffle` and `exchange` terminology interchangeably ? IIRC, this was the case on the original design docs. 

So when we say `shuffle` is disabled, does that mean shuffle based exchange is disabled (but broadcast exchange can happen) or `exchange` all together is disabled ?

**Comment by 61yao** [comment]
> I am guessing we can use this to compare performance and even correctness between `broadcast join` , `shuffle join`, `colocated partitioning` table join ?
> 
> I guess pretty much any scenario where we force the plan to not have an exchange ?
> 

SkipShuffle is mostly used for doing colocated table join today. The main purpose is to compare colocated partitioning join and not.  

> Another question - are we using `shuffle` and `exchange` terminology interchangeably ? IIRC, this was the case on the original design docs.
> 
> So when we say `shuffle` is disabled, does that mean shuffle based exchange is disabled (but broadcast exchange can happen) or `exchange` all together is disabled ?

Disable is for disable SkipShuffle optimizer. SkipShuffle here means we are still doing exchange. but we are changing the broadcast exchange to singleton to favor local exchange. (we are already seeing problems caused by SkipShuffle and considering removing it all together )

**Comment by siddharthteotia** [comment]
> > I am guessing we can use this to compare performance and even correctness between `broadcast join` , `shuffle join`, `colocated partitioning` table join ?
> > I guess pretty much any scenario where we force the plan to not have an exchange ?
> 
> SkipShuffle is mostly used for doing colocated table join today. The main purpose is to compare colocated partitioning join and not.
> 
> > Another question - are we using `shuffle` and `exchange` terminology interchangeably ? IIRC, this was the case on the original design docs.
> > So when we say `shuffle` is disabled, does that mean shuffle based exchange is disabled (but broadcast exchange can happen) or `exchange` all together is disabled ?
> 
> Disable is for disable SkipShuffle optimizer. SkipShuffle here means we are still doing exchange. but we are changing the broadcast exchange to singleton to favor local exchange. (we are already seeing problems caused by SkipShuffle and considering removing it all together )

Got it !! Makes sense

**Comment by siddharthteotia** [comment]
May want to update the naming as discussed offline to avoid confusion. Otherwise, this is good.

**Comment by 61yao** [comment]
> May want to update the naming as discussed offline to avoid confusion. Otherwise, this is good.

Renamed the option. Thanks for the suggestion!

**Review by siddharthteotia — APPROVED** [review]

**Comment by walterddr** [comment]
hmm why is this necessary? i thought we established the fact that shuffle rewrite should not be used going forward.

**Comment by 61yao** [comment]
> hmm why is this necessary? i thought we established the fact that shuffle rewrite should not be used going forward.

Because we don't want to wait till that API change to happen.especially we are doing a bunch of perf testing now. it would be nice to have an option to disable this since we know it is problematic. Or I can remove all skipShuffle code now if needed.

**Comment by walterddr** [comment]
let's remove all skipShuffle usage for now. (keep the shuffle rewrite, some of the logics are valid. just the modification to singleton hash is not desired)

**Comment by 61yao** [comment]
> let's remove all skipShuffle usage for now. (keep the shuffle rewrite, some of the logics are valid. just the modification to singleton exchange is not desired)

How about I change the default behavior to be disabled and enabled only requested? So we can use it when we think it is correct and understand the perf gain by this?

**Review by walterddr — APPROVED** [review]

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/StagePlanner.java [hunk 1]
```diff
@@ -65,8 +65,6 @@ public QueryPlan makePlan(RelRoot relRoot) {
 
     // walk the plan and create stages.
     StageNode globalStageRoot = walkRelPlan(relRootNode, getNewStageId());
-    ShuffleRewriteVisitor.optimizeShuffles(globalStageRoot);
-
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/StagePlanner.java:69]: with this you will also have to disable the shuffle rewrite test in planner

**Pull request merged**